File: app/src/main/java/com/p1/uber/activities/client/DetailRequestActivity.java
package com.p1.uber.activities.client;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.JointType;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.android.gms.maps.model.SquareCap;
import com.p1.uber.R;
import com.p1.uber.include.MyToolbar;
import com.p1.uber.provides.GoogleApiProvider;
import com.p1.uber.utils.DecodePoints;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DetailRequestActivity extends AppCompatActivity implements OnMapReadyCallback {

    private GoogleMap mMap;
    private SupportMapFragment mMapFragment;

    private double mExtraOriginLat;
    private double mExtraOriginLng;
    private double mExtraDestinationLat;
    private double mExtraDestinationLng;

    private String mExtraOrigin;
    private String mExtraDestination;

    private LatLng mOriginLatLng;
    private LatLng mDestinationLatLng;

    private GoogleApiProvider mGoogleApiProvider;

    private List<LatLng> mPolylineList;
    private PolylineOptions mPolylineOptions;

    private TextView mTextViewOrigin;
    private TextView mTextViewDestination;
    private TextView mTextViewTime;
    private TextView mTextViewDistance;

    private Button mButtonRequest;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail_request);
        MyToolbar.show(this,"TUS DATOS",true);
        mMapFragment= (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
        mMapFragment.getMapAsync(this);

        mExtraOriginLat = getIntent().getDoubleExtra("origin_lat",0);
        mExtraOriginLng = getIntent().getDoubleExtra("origin_lng",0);
        mExtraDestinationLat = getIntent().getDoubleExtra("destination_lat",0);
        mExtraDestinationLng = getIntent().getDoubleExtra("destination_lng",0);

        mOriginLatLng = new LatLng(mExtraOriginLat,mExtraOriginLng);
        mDestinationLatLng = new LatLng(mExtraDestinationLat,mExtraDestinationLng);

        mGoogleApiProvider = new GoogleApiProvider(DetailRequestActivity.this);

        mTextViewOrigin = findViewById(R.id.textViewOrigin);
        mTextViewDestination = findViewById(R.id.textViewDestination);
        mTextViewTime = findViewById(R.id.textViewTime);
        mTextViewDistance = findViewById(R.id.textViewDistance);

        mExtraOrigin = getIntent().getStringExtra("origin");
        mExtraDestination = getIntent().getStringExtra("destination");

        mButtonRequest = findViewById(R.id.btnRequestNow);


        mTextViewOrigin.setText(mExtraOrigin);
        mTextViewDestination.setText(mExtraDestination);

        mButtonRequest.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                goToRequestDiver();
            }
        });



    }

    private void goToRequestDiver(){
        Intent intent = new Intent(DetailRequestActivity.this, RequestDriverActivity.class);
        intent.putExtra("origin_lat",mOriginLatLng.latitude);
        intent.putExtra("origin_lng",mOriginLatLng.longitude);
        intent.putExtra("origin",mExtraOrigin);
        intent.putExtra("destination",mExtraDestination);
        intent.putExtra("destination_lat",mDestinationLatLng.latitude);
        intent.putExtra("destination_lng",mDestinationLatLng.longitude);
        startActivity(intent);
        finish();
    }


    private void drawRoute(){
        mGoogleApiProvider.getDirections(mOriginLatLng, mDestinationLatLng).enqueue(new Callback<String>() {
            @Override
            public void onResponse(Call<String> call, Response<String> response) {
                try {

                    JSONObject jsonObject = new JSONObject(response.body());
                    JSONArray jsonArray =  jsonObject.getJSONArray("routes");
                    JSONObject route = jsonArray.getJSONObject(0);
                    JSONObject polylines = route.getJSONObject("overview_polyline");
                    String points = polylines.getString("points");
                    mPolylineList = DecodePoints.decodePoly(points);
                    mPolylineOptions = new PolylineOptions();
                    mPolylineOptions.color(Color.DKGRAY);
                    mPolylineOptions.width(8f);
                    mPolylineOptions.startCap(new SquareCap());
                    mPolylineOptions.jointType(JointType.ROUND);
                    mPolylineOptions.addAll(mPolylineList);
                    mMap.addPolyline(mPolylineOptions);

                    JSONArray legs = route.getJSONArray("legs");
                    JSONObject leg = legs.getJSONObject(0);
                    JSONObject distance = leg.getJSONObject("distance");
                    JSONObject duration = leg.getJSONObject("duration");
                    String distanceText = distance.getString("text");
                    String durationText = duration.getString("text");
                    mTextViewTime.setText(durationText);
                    mTextViewDistance.setText(distanceText);



                } catch (Exception e) {
                    Log.d("Error", "Error encontrado" + e.getMessage());
                }
            }

            @Override
            public void onFailure(Call<String> call, Throwable t) {

            }
        });
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap=googleMap;
        mMap.setMapType(GoogleMap.MAP_TYPE_NORMAL);
        mMap.getUiSettings().setZoomControlsEnabled(true);

        mMap.addMarker(new MarkerOptions().position(mOriginLatLng).title("Origen").icon(BitmapDescriptorFactory.fromResource(R.drawable.icons8_mapa_de_pin_100)));
        mMap.addMarker(new MarkerOptions().position(mDestinationLatLng).title("Destino").icon(BitmapDescriptorFactory.fromResource(R.drawable.icons8_mapa_de_pin)));
        mMap.animateCamera(CameraUpdateFactory.newCameraPosition(
                new CameraPosition.Builder().target(mOriginLatLng)
                .zoom(14f).build()
        ));

        drawRoute();
    }
}

File: app/src/main/java/com/p1/uberfares/activities/client/DetailRequestActivity.kt
package com.p1.uberfares.activities.client

import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.Button
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.JointType
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MapStyleOptions
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.PolylineOptions
import com.google.android.gms.maps.model.SquareCap
import com.p1.uberfares.R
import com.p1.uberfares.provides.GoogleApiProvider
import com.p1.uberfares.utils.DecodePoints.decodePoly
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class DetailRequestActivity : AppCompatActivity(), OnMapReadyCallback {
    private var mMap: GoogleMap? = null
    private var mMapFragment: SupportMapFragment? = null
    private var mExtraOriginLat = 0.0
    private var mExtraOriginLng = 0.0
    private var mExtraDestinationLat = 0.0
    private var mExtraDestinationLng = 0.0
    private var mExtraOrigin: String? = null
    private var mExtraDestination: String? = null
    private var mOriginLatLng: LatLng? = null
    private var mDestinationLatLng: LatLng? = null
    private var type: String? = null
    private var mGoogleApiProvider: GoogleApiProvider? = null
    private var mPolylineList: List<LatLng>? = null
    private var mPolylineOptions: PolylineOptions? = null

    private lateinit var mRlRechargeAlipay: ConstraintLayout
    private lateinit var mRlRechargeWechat: ConstraintLayout
    private lateinit var mButtonRequest: Button

    @RequiresApi(Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail_request)
        mMapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
        mMapFragment!!.getMapAsync(this)
        mExtraOriginLat = intent.getDoubleExtra("origin_lat", 0.0)
        mExtraOriginLng = intent.getDoubleExtra("origin_lng", 0.0)
        mExtraDestinationLat = intent.getDoubleExtra("destination_lat", 0.0)
        mExtraDestinationLng = intent.getDoubleExtra("destination_lng", 0.0)
        mOriginLatLng = LatLng(mExtraOriginLat, mExtraOriginLng)
        mDestinationLatLng = LatLng(mExtraDestinationLat, mExtraDestinationLng)
        mGoogleApiProvider = GoogleApiProvider(this@DetailRequestActivity)
        // mTextViewDestination = findViewById(R.id.textViewDestination)
        mExtraOrigin = intent.getStringExtra("origin")
        mExtraDestination = intent.getStringExtra("destination")
        mButtonRequest = findViewById(R.id.btnRequestNow)
//        mRgRecharge = findViewById(R.id.mRgRecharge)
        type = "Motorcycle"
        mRlRechargeWechat = findViewById(R.id.mRlRechargeWechat)
        mRlRechargeAlipay = findViewById(R.id.mRlRechargeAlipay)
        val myLayout = findViewById<LinearLayout>(R.id.linearLayout3)
        myLayout.visibility = View.VISIBLE
        val animation: Animation = AnimationUtils.loadAnimation(this, R.anim.bottom_up_animation)
        myLayout.startAnimation(animation)
        mRlRechargeWechat.setOnClickListener {

            val colorStateList =
                ColorStateList.valueOf(getColor(R.color.colorAccent)) // تعيين لون الخلفية هنا (في هذا المثال اللون الأحمر)
            mRlRechargeWechat.setBackgroundTintList(colorStateList)

            val colorStateList2 =
                ColorStateList.valueOf(getColor(R.color.d6d6d6)) // تعيين لون الخلفية هنا (في هذا المثال اللون الأحمر)
            mRlRechargeAlipay.setBackgroundTintList(colorStateList2)
            type = "Car"
        }

        mRlRechargeAlipay.setOnClickListener {

            val colorStateList12 =
                ColorStateList.valueOf(getColor(R.color.d6d6d6)) // تعيين لون الخلفية هنا (في هذا المثال اللون الأحمر)
            mRlRechargeWechat.setBackgroundTintList(colorStateList12)

            val colorStateList13 =
                ColorStateList.valueOf(getColor(R.color.colorAccent)) // تعيين لون الخلفية هنا (في هذا المثال اللون الأحمر)
            mRlRechargeAlipay.setBackgroundTintList(colorStateList13)
            type = "Motorcycle"
        }




        mButtonRequest.setOnClickListener(
            { goToRequestDiver() }
        )

    }

    private fun goToRequestDiver() {
        val intent = Intent(this@DetailRequestActivity, RequestDriverActivity::class.java)
        intent.putExtra("origin_lat", mOriginLatLng!!.latitude)
        intent.putExtra("origin_lng", mOriginLatLng!!.longitude)
        intent.putExtra("origin", mExtraOrigin)
        intent.putExtra("destination", mExtraDestination)
        intent.putExtra("destination_lat", mDestinationLatLng!!.latitude)
        intent.putExtra("destination_lng", mDestinationLatLng!!.longitude)
        intent.putExtra("type", type)
        startActivity(intent)
        finish()
    }

    private fun drawRoute() {
        mGoogleApiProvider!!.getDirections(mOriginLatLng!!, mDestinationLatLng!!)!!
            .enqueue(object : Callback<String?> {
                override fun onResponse(call: Call<String?>, response: Response<String?>) {
                    try {
                        val jsonObject = JSONObject(response.body())
                        val jsonArray = jsonObject.getJSONArray("routes")
                        val route = jsonArray.getJSONObject(0)
                        val polylines = route.getJSONObject("overview_polyline")
                        val points = polylines.getString("points")
                        mPolylineList = decodePoly(points)
                        mPolylineOptions = PolylineOptions()
                        mPolylineOptions!!.color(Color.DKGRAY)
                        mPolylineOptions!!.width(8f)
                        mPolylineOptions!!.startCap(SquareCap())
                        mPolylineOptions!!.jointType(JointType.ROUND)
                        mPolylineOptions!!.addAll(mPolylineList!!)
                        mMap!!.addPolyline(mPolylineOptions!!)
                        val legs = route.getJSONArray("legs")
                        val leg = legs.getJSONObject(0)
                        val distance = leg.getJSONObject("distance")
                        val duration = leg.getJSONObject("duration")
                        val distanceText = distance.getString("text")
                        val durationText = duration.getString("text")
                    } catch (e: Exception) {
                        Log.d("Error", "bug found" + e.message)
                    }
                }

                override fun onFailure(call: Call<String?>, t: Throwable) {}
            })
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        mMap!!.mapType = GoogleMap.MAP_TYPE_NORMAL
        mMap!!.uiSettings.isZoomControlsEnabled = true
        val styleOptions = MapStyleOptions.loadRawResourceStyle(this, R.raw.map_stule)
        mMap!!.setMapStyle(styleOptions)
        mMap!!.addMarker(
            MarkerOptions().position(mOriginLatLng!!).title("Origen")
                .icon(BitmapDescriptorFactory.fromResource(R.drawable.icons8_mapa_de_pin_100))
        )
        mMap!!.addMarker(
            MarkerOptions().position(mDestinationLatLng!!).title("Destino")
                .icon(BitmapDescriptorFactory.fromResource(R.drawable.icons8_mapa_de_pin))
        )
        mMap!!.animateCamera(
            CameraUpdateFactory.newCameraPosition(
                CameraPosition.Builder().target(mOriginLatLng!!)
                    .zoom(14f).build()
            )
        )
        drawRoute()
    }

}